### ISSUE TITLE
Crash when attempting to open a .bin file

### ISSUE THREAD

**Issue by bingbingyahoo27** [MM102/SNES-Tools-for-Aseprite#1]
When running the SNES tools Lua script for Aseprite, after trying to open a .bin file, the editor crashes.

1. Click on import .bin
2. Select a .bin file, which was outputted from Lunar Magic
3. When I open it, the editor crashes.

Unfortunately, the script debugging is really not that good on Aseprite. Best I can do is leave a basic stack trace from what I can see on the debugger:

This is the trace after I click on the file in the file browser when trying to import as a new sprite -
1. SNES_GFX_tools.lua:194: app.transaction(
2. [string "internal"]:1: SNES_GFX_tools.lua:194: app.transaction(
3. SNES_GFX_tools.lua:283: end)
4. SNES_GFX_tools.lua:194: app.transaction(
5. Aseprite then proceeds to crash

I just updated to the latest stable version of Aseprite (1.2.33), and I'm using the initial release that was on the main branch. The download provided in the video description didn't work for me either, unfortunately. I've tried it with linked and unlinked bin files.

This tool is probably really nice and severely underrated, but it's not working for me.

**Comment by MM102** [comment]
Can you provide the .bin file you were working with? I'm having trouble reproducing this

**Comment by bingbingyahoo27** [comment]
Sure thing. Here are a couple .bin files that don't work for me, with varying sizes: https://github.com/bingbingyahoo27/gfx-files

They're all just vanilla .bin files extracted from Lunar Magic, from a SMW (USA) ROM.

**Comment by bingbingyahoo27** [comment]
Whoops, forgot to mention - I'm working on Windows 11, 64 bit.

**Comment by MM102** [comment]
hmm I can't seem to reproduce this.
can you try this version whre I've removed the app.transaction lines to see if you get a different error?
[no-transactions.zip](https://github.com/MM102/SNES-Tools-for-Aseprite/files/8209572/no-transactions.zip)

**Comment by bingbingyahoo27** [comment]
Yep, I got a different error this time. With those lines removed, after opening the file from the file browser, I get the error message `attempt to index a function value` from a pop-up console window. No crash, either - the "Import .bin" button doesn't seem to work at all after this if I click it, probably due to the error being handled.

**Comment by bingbingyahoo27** [comment]
Whoops, here's the debugger trace for the transactionless script. I renamed the transactionless script file for convenience:

1. Enable the debugger, and open a .bin file with the script.
2. ... [string "internal"]:1: SNES_gfx_tools-test-transactionless.lua:194: `if DLG_import.data.importfile == "Import" then`
3. ... [string "internal"]:1: SNES_gfx_tools-test-transactionless.lua:197: `if not app.fs.isFile(DLG_import.data.importfile) then`
4. ... [string "internal"]:1: SNES_gfx_tools-test-transactionless.lua:201: `if not app.fs.fileExtension(DLG_import.data.importfile) == "bin" then`
5. ... [string "internal"]:1: `SNES_gfx_tools-test-transactionless.lua:205: local input = io.open(app.fs.normalizePath(DLG_import.data.importfile), "rb")`
6. `attempt to index a function value`

Hope this helps - let me know if you need any more info. Thanks!

**Comment by bingbingyahoo27** [comment]
So I looked into it, and it looks like everything functions as expected until the actual file import. I felt like it's something OS-specific with file paths that's not being parsed correctly, somehow, but no, that doesn't actually seem to be the case.

What's weird is that if I add the following line to display the path it's trying to open:
SNES_gfx_tools-test-transactionless.lua:205: `app.alert(app.fs.normalizePath(DLG_import.data.importfile))`

My path looks totally fine. It's well formatted, and when I type it into the file explorer it works fine.

I tried opening the file without opening it as binary (ie. `io.open(app.fs.normalizePath(DLG_import.data.importfile), "r")` and I get the same error*. Hmm.

**Comment by bingbingyahoo27** [comment]
Do you know of any workarounds I could try in the meantime?

**Comment by MM102** [comment]
just to be sure, app.fs.normalizePath outputs the path with backslashes right?
also, you can try removing the app.fs.normalizePath part entirely
so
`io.open(app.fs.normalizePath(DLG_import.data.importfile), "r")`
becomes
`io.open(DLG_import.data.importfile, "r")`

**Comment by bingbingyahoo27** [comment]
Apologies for the delay. Yes, looks like from the alert I added, the path is formatted with backslashes.

I removed the path normalization, and it looks like the path is still the same as before; same error.

Maybe this is only a problem for opening these specific types of binaries? I might check around for other lua scripts and see if I can find one that opens an image, to test if it's also broken.
